Comment symbol in header support for CSV reader

hermanndetz opened the pull request:
The main branch simply parses the header row into the column names, if it is identified to be a header. This has the effect that for CSV files with comments, the comment symbol becomes part of the first column name.

Take e.g. the file header:
# Timestamp	Setpoint	Measured

The name of the first column would be "# Timestamp", which makes it hard to use in SQL statements. The modifications in the advcsv branch allow to define a list of comment symbols that currently only contains'#', which are ignored in this situation. The resulting column name of the first column would then simply be "Timestamp".

hermanndetz committed:
Ignoring comment symbol in header row

Added the ability to define a list of comment symbols (like the commonly
used '#') that are ignored, when parsing the column names from the
header row.

hermanndetz committed:
Format fixes for previous commit

hannesmuehleisen commented:
@tdoehmen  could you have a look please?

Mytherin commented:
Could you also add some test cases under `test/sql/copy/csv` that verify that this behavior works as expected?

hermanndetz committed:
Another fix to propoerly clean column name

After removing the comment sign, also the following whitespace or column
separator is removed properly.

hermanndetz commented:
> Thanks for the PR! Could you add some test cases under `test/sql/copy/csv` that verify that this behavior works as expected?

Could you direct me, how to write a test that checks for the header row of a table? I'm trying since days without success. Or point me to an existing test case that does similar things. Couldn't find one, but maybe I missed it.

Mytherin commented:
Are you interested in the column names specifically, or just how to write a test in general?

For a test in general check e.g. [test/sql/copy/csv/test_nfc.test](https://github.com/duckdb/duckdb/blob/master/test/sql/copy/csv/test_nfc.test), which reads [test/sql/copy/csv/data/test/nfc.csv](https://github.com/duckdb/duckdb/blob/6c7c9805fdf1604039ebed47d233ea55cabb4b2c/test/sql/copy/csv/data/test/nfc.csv).

If you want to verify the column names, you could just write a query that checks e.g. whether or not the column exists (`select my_name from read_csv(...)`), or get a list of all column names in a table using a query:

```sql
create table integers(i integer, j integer);
select name from pragma_table_info('integers');
┌──────┐
│ name │
├──────┤
│ i    │
│ j    │
└──────┘
```

hermanndetz committed:
Checking for comment symbol only in first CSV row

hermanndetz committed:
Added tests for header parsing in CSV files

hermanndetz commented:
> Are you interested in the column names specifically, or just how to write a test in general?
> 
> For a test in general check e.g. [test/sql/copy/csv/test_nfc.test](https://github.com/duckdb/duckdb/blob/master/test/sql/copy/csv/test_nfc.test), which reads [test/sql/copy/csv/data/test/nfc.csv](https://github.com/duckdb/duckdb/blob/6c7c9805fdf1604039ebed47d233ea55cabb4b2c/test/sql/copy/csv/data/test/nfc.csv).
> 
> If you want to verify the column names, you could just write a query that checks e.g. whether or not the column exists (`select my_name from read_csv(...)`), or get a list of all column names in a table using a query:
> 
> ```sql
> create table integers(i integer, j integer);
> select name from pragma_table_info('integers');
> ┌──────┐
> │ name │
> ├──────┤
> │ i    │
> │ j    │
> └──────┘
> ```

Thanks a lot! I added some tests now

hermanndetz commented on src/execution/operator/persistent/buffered_csv_reader.cpp:
```diff
@@ -679,18 +679,19 @@ vector<LogicalType> BufferedCSVReader::SniffCSV(const vector<LogicalType> &reque
-                }
-            }
+			// check if beginning of col name matches beginning of a comment
+			for (auto comment_candidate : comment_candidates) {
```
Changed now, so only the first column is examined for comment symbols.

Mytherin reviewed (approved):
Thanks, looks good to me now.

hermanndetz commented on src/execution/operator/persistent/buffered_csv_reader.cpp:
```diff
@@ -679,18 +679,19 @@ vector<LogicalType> BufferedCSVReader::SniffCSV(const vector<LogicalType> &reque
-                }
-            }
+			// check if beginning of col name matches beginning of a comment
+			for (auto comment_candidate : comment_candidates) {
```
Done.

hermanndetz committed:
Format fixes for previous commit

hermanndetz commented:
> Thanks, looks good to me now.

There's one conceptional problem related to quoted column names: The format of the CSV file is guessed within the SniffCSV function, which gets data that are read by the ParseCSV function. The ParseCSV function removes the quotes, therefore, the SniffCSV function at the moment does not know, if e.g. "# Time" was within quotes originally and removes the '#' no matter what.

Two possible options here:
* Go for a more complex solution to account for that. This would actually first require to define a desired behavior.
* Document the behavior as is

tdoehmen commented:
It would be possible to find out whether the field was quoted or not by tracking what happens in the state machine which parses the CSV (`ParseSimpleCSV` or  `ParseComplexCSV`). But this could be overkill only for solving this problem. 

Though, even if it wasn't quoted, the behavior might still be a little unexpected, as in the following case:

`#ID,#Target,#Name`

Would parse to column names being 

"_ID_", "_#Target_" and "_#Name_"

My suggestion would be to simply add an optional parameter, called something like `skip_comment_in_header`, to the `read_csv(_auto)` and `COPY INTO` function which activates the implemented behavior, and documenting it [here](https://github.com/duckdb/duckdb-web/blob/master/docs/sql/statements/copy.md). I can also help you out with adding the parameter if you like.

hermanndetz commented:
> It would be possible to find out whether the field was quoted or not by tracking what happens in the state machine which parses the CSV (`ParseSimpleCSV` or `ParseComplexCSV`). But this could be overkill only for solving this problem.
> 
> Though, even if it wasn't quoted, the behavior might still be a little unexpected, as in the following case:
> 
> `#ID,#Target,#Name`
> 
> Would parse to column names being
> 
> "_ID_", "_#Target_" and "_#Name_"
> 
> My suggestion would be to simply add an optional parameter, called something like `skip_comment_in_header`, to the `read_csv(_auto)` and `COPY INTO` function which activates the implemented behavior, and documenting it [here](https://github.com/duckdb/duckdb-web/blob/master/docs/sql/statements/copy.md). I can also help you out with adding the parameter if you like.

I'd be fine with that solution. Don't yet have the big picture, so I don't know, if it would break something.

hawkfish commented:
It is worth noting that [RFC4180](https://datatracker.ietf.org/doc/html/rfc4180) does not define comment lines characters at all and Python [rejected](https://bugs.python.org/issue1225769) this functionality. That said, it seems like a strange situation here where a line that is commented out is assumed to contain information *after* the comment?

Maybe a better solution is to have some mechanism for reading CSVs from a Python file-like object and then users can [filter and massage it however they want](https://pypi.org/project/textform/0.8.0/)?

hermanndetz commented:
I'm aware that the comments are not standardized for this file format. Yet, reality seem to partially deviate from the standard here. With the mixture of CSV dialects that are out there, it's probably impossible to come up with a solution that fits everybody.

How about leaving a strict mode as default, which sticks to the standard, but with some option to account for commented header rows or also as option to remove bad data points. I know that the later can be done in software, but again, I'm thinking of loads of data that already exist in that format.

tdoehmen commented:
@hawkfish Nice! That seems like an interesting tool, thanks for sharing! In a sense, we are already deviating from RFC-4180 because we support other separators and quoting styles that appear "in the wild". So, if comments "are a thing" I wouldn't say that it's generally a bad idea to add optional support for it. But I agree that "commented headers" are again another pair of shoes.

@hermanndetz I feel like - correct me if I'm wrong - your main pain point here is that the comment symbol (e.g. "# ") will end up in the column name, which makes it hard to use in SQL queries. Wouldn't some kind of general normalization procedure of column names (lowercase, removing spaces, symbols etc...) solve that problem as well? We once decided to keep the column names as they appear in the input file, but I think having an _option_ to just normalize all the column names, so they are easy to use in an SQL query would have a great general utility. 

Removing "junk" (comments, padding lines/columns etc.) from the input then feels like a whole different topic. We try to automatically skip junk at the beginning of the file (before the header) at the moment. Skipping junk at the end could also be done, but catching junk somewhere in the middle seems like quite a challenge (without some knowledge like "# is a comment symbol"). And I think in such a case using a tool like _textform_ would be the better way to go, because there are so many different ways in which CSV files can be "misused" and lead to junk rows/columns (just think about Excel-Spreadsheet exports to CSV).

Mytherin commented:
> @hermanndetz I feel like - correct me if I'm wrong - your main pain point here is that the comment symbol (e.g. "# ") will end up in the column name, which makes it hard to use in SQL queries. Wouldn't some kind of general normalization procedure of column names (lowercase, removing spaces, symbols etc...) solve that problem as well? We once decided to keep the column names as they appear in the input file, but I think it would be great to have an _option_ to just normalize all the column names, so they are easy to use in an SQL query.
> 
> Removing "junk" (comments, padding lines/columns etc.) from the input then feels like a whole different topic. We try to automatically skip junk at the beginning of the file (before the header) at the moment. Skipping junk at the end could also be done, but catching junk somewhere in the middle seems like quite a challenge (without some knowledge like "# is a comment symbol"). And I think in such a case using a tool like _textform_ would be the better way to go, because there are so many different ways in which CSV files can be "misused" and lead to junk rows/columns (just think about Excel-Spreadsheet exports to CSV) that we can't possibly solve all of these with CSV parser options.

I completely agree that this solution seems like a better solution to the problem at hand. The current PR has the side effect that information from the headers is lost, which may work as expected for certain CSV files but cause surprising behavior when parsing others. Adding a flag specifically to handle this one scenario also seems excessive.

Adding a flag such as "`normalize_names`" seems much more generally useful, as it would solve this problem for all complex column names, such as column names containing spaces or other weird characters.

hermanndetz commented:
> > @hermanndetz I feel like - correct me if I'm wrong - your main pain point here is that the comment symbol (e.g. "# ") will end up in the column name, which makes it hard to use in SQL queries. Wouldn't some kind of general normalization procedure of column names (lowercase, removing spaces, symbols etc...) solve that problem as well? We once decided to keep the column names as they appear in the input file, but I think it would be great to have an _option_ to just normalize all the column names, so they are easy to use in an SQL query.
> > Removing "junk" (comments, padding lines/columns etc.) from the input then feels like a whole different topic. We try to automatically skip junk at the beginning of the file (before the header) at the moment. Skipping junk at the end could also be done, but catching junk somewhere in the middle seems like quite a challenge (without some knowledge like "# is a comment symbol"). And I think in such a case using a tool like _textform_ would be the better way to go, because there are so many different ways in which CSV files can be "misused" and lead to junk rows/columns (just think about Excel-Spreadsheet exports to CSV) that we can't possibly solve all of these with CSV parser options.
> 
> I completely agree that this solution seems like a better solution to the problem at hand. The current PR has the side effect that information from the headers is lost, which may work as expected for certain CSV files but cause surprising behavior when parsing others. Adding a flag specifically to handle this one scenario also seems excessive.
> 
> Adding a flag such as "`normalize_names`" seems much more generally useful, as it would solve this problem for all complex column names, such as column names containing spaces or other weird characters.

Fully agree here. Sounds like the best solution.

Mytherin commented:
This PR has been superseded by #1875. Closing this. Thanks for the effort & discussion!
